torch/optim/lr_scheduler.py

```diff
@@ -0,0 +1,206 @@
+    def __init__(self, optimizer, base_lrs, lr_lambdas):
+        self.optimizer = optimizer
+        self.base_lrs = base_lrs
+        self.lr_lambdas = lr_lambdas
+
+    def step(self, epoch):
+        for param_group, base_lr, lr_lambda in zip(
+                self.optimizer.param_groups,
+                self.base_lrs, self.lr_lambdas):
+            param_group['lr'] = base_lr * lr_lambda(epoch)
+
+
+class StepLR(LambdaLR):
+    """Set the learning rate to the base_lr decayed by gamma
+    every step_size epochs.
+
+
+    Example:
+        >>> # lr = 0.05     if epoch < 30
+        >>> # lr = 0.005    if 30 <= epoch < 60
+        >>> # lr = 0.0005   if 60 <= epoch < 90
+        >>> # ...
+        >>> scheduler = StepLR(optimizer, base_lr=0.05, gamma=0.1, step_size=30)
+        >>> for epoch in range(100):
+        >>>     scheduler.step(epoch)
+        >>>     train(...)
+        >>>     validate(...)
+    """
+
+    def __init__(self, optimizer, base_lr=0.1, gamma=0.1, step_size=30):
+        super(StepLR, self).__init__(optimizer, base_lr,
+                                     lambda epoch: gamma ** (epoch // step_size))
+
+
+class MultiStepLR(LambdaLR):
+    """Set the learning rate to the base_lr decayed by gamma
+    once the number of epoch reaches one of the milestones.
+
+
+    Example:
+        >>> # lr = 0.05     if epoch < 30
+        >>> # lr = 0.005    if 30 <= epoch < 80
+        >>> # lr = 0.0005   if epoch >=80
+        >>> scheduler = MultiStepLR(optimizer, base_lr=0.05, gamma=0.1, milestones=[30,80])
+        >>> for epoch in range(100):
+        >>>     scheduler.step(epoch)
+        >>>     train(...)
+        >>>     validate(...)
+    """
+
+    def __init__(self, optimizer, base_lr=0.1, gamma=0.1, milestones=(10, 20, 30)):
+        milestones = sorted(milestones)
+        super(MultiStepLR, self).__init__(optimizer, base_lr,
+                                          lambda epoch: gamma ** bisect_right(milestones, epoch))
+
+
+class ExponentialLR(LambdaLR):
+    """Set the learning rate to the initial LR decayed by gamma in
+    every epoch."""
+
+    def __init__(self, optimizer, base_lr, gamma):
+        super(ExponentialLR, self).__init__(optimizer, base_lr,
+                                            lambda epoch: gamma ** epoch)
+
+
+class ReduceLROnPlateau(object):
+    """Reduce learning rate when a metric has stopped improving.
+    Models often benefit from reducing the learning rate by a factor
+    of 2-10 once learning stagnates. This scheduler reads a metrics
+    quantity and if no improvement is seen for a 'patience' number
+    of epochs, the learning rate is reduced.
+
+    Args:
+        factor: factor by which the learning rate will
+            be reduced. new_lr = lr * factor
+        patience: number of epochs with no improvement
+            after which learning rate will be reduced.
+        verbose: int. 0: quiet, 1: update messages.
+        mode: one of {min, max}. In `min` mode,
```

Comment:
I am not quite understanding this comment.  Max returns a value instead of a boolean. Also, making it an option would make it easier to understand.